src/meta/src/barrier/info.rs

```diff
@@ -12,57 +12,139 @@
 
-/// [`BarrierActorInfo`] resolves the actor info read from meta store for
+#[derive(Debug, Clone)]
+pub struct ActorDesc {
+    pub id: ActorId,
+    pub node_id: WorkerId,
+    pub is_injectable: bool,
+}
+
+#[derive(Debug, Clone)]
+pub enum CommandActorChanges {
+    Actor {
+        to_add: Vec<ActorDesc>,
+        to_remove: HashSet<ActorId>,
+    },
+    None,
+}
+
+/// [`InflightActorInfo`] resolves the actor info read from meta store for
 /// [`crate::barrier::GlobalBarrierManager`].
-pub struct BarrierActorInfo {
+#[derive(Default, Clone)]
+pub struct InflightActorInfo {
     /// node_id => node
-    pub node_map: HashMap<WorkerId, WorkerNode>,
+    pub node_map: HashMap<WorkerId, PbWorkerNode>,
 
     /// node_id => actors
-    pub actor_map: HashMap<WorkerId, Vec<ActorId>>,
+    pub actor_map: HashMap<WorkerId, HashSet<ActorId>>,
 
     /// node_id => barrier inject actors
-    pub actor_map_to_send: HashMap<WorkerId, Vec<ActorId>>,
+    pub actor_map_to_send: HashMap<WorkerId, HashSet<ActorId>>,
+
+    /// actor_id => WorkerId
+    pub actor_location_map: HashMap<ActorId, WorkerId>,
 }
 
-impl BarrierActorInfo {
-    // TODO: we may resolve this info as graph updating, instead of doing it every time we want to
-    // send a barrier
+impl InflightActorInfo {
     pub fn resolve(
-        all_nodes: impl IntoIterator<Item = WorkerNode>,
+        all_nodes: impl IntoIterator<Item = PbWorkerNode>,
         actor_infos: ActorInfos,
     ) -> Self {
         let node_map = all_nodes
             .into_iter()
             .map(|node| (node.id, node))
             .collect::<HashMap<_, _>>();
 
+        let actor_map = actor_infos
+            .actor_maps
+            .into_iter()
+            .map(|(node_id, actor_ids)| (node_id, actor_ids.into_iter().collect::<HashSet<_>>()))
+            .collect::<HashMap<_, _>>();
+
+        let actor_map_to_send = actor_infos
+            .barrier_inject_actor_maps
+            .into_iter()
+            .map(|(node_id, actor_ids)| (node_id, actor_ids.into_iter().collect::<HashSet<_>>()))
+            .collect::<HashMap<_, _>>();
+
+        let actor_location_map = actor_map
+            .iter()
+            .flat_map(|(node_id, actor_ids)| actor_ids.iter().map(|actor_id| (*actor_id, *node_id)))
+            .collect::<HashMap<_, _>>();
+
         Self {
             node_map,
-            actor_map: actor_infos.actor_maps,
-            actor_map_to_send: actor_infos.barrier_inject_actor_maps,
+            actor_map,
+            actor_map_to_send,
+            actor_location_map,
+        }
+    }
+
+    pub fn resolve_worker_nodes(&mut self, all_nodes: impl IntoIterator<Item = PbWorkerNode>) {
+        self.node_map = all_nodes
+            .into_iter()
+            .map(|node| (node.id, node))
+            .collect::<HashMap<_, _>>();
+    }
+
+    pub fn pre_apply(&mut self, changes: &CommandActorChanges) {
+        if let CommandActorChanges::Actor { to_add, .. } = changes {
```

Comment:
Will there be other variants? If nope, I would suggest directly using `Option` here or this looks concerning. 😄